### DISCUSSION TITLE
Extend padding-line-between-statements StatementTypes

### DISCUSSION THREAD

**Discussion by maluramichael** [eslint/eslint#14844] [Q&A]
**What rule do you want to change?**
padding-line-between-statements

**Does this change cause the rule to produce more or fewer warnings?**
Potentially more

**How will the change be implemented? (New option, new default behavior, etc.)?**
By extending the `StatementTypes` object with a new test.

**Please provide some example code that this change will affect:**

There is a code style rule in our company that there has to be a space between every variable declaration followed by a function call and vice versa.

```js
const x = 10;
const y = 20;
const result = add(x, y); // <- thats ok because it is also a variable declaration
const p = 30;
```

```js
const x = 10;
const y = 20;
printAdd(x, y); // <- this would be wrong because the declaration of y is followed by a call to printAdd
const p = 30; // <- would be also wrong
```

**What does the rule currently do for this code?**
Right now nothing because i don't think i can configure the rule to my needs without extending the StatementTypes object.

**What will the rule do after it's changed?**
Adding a space between assignments and function calls.

```js
const x = 10;
const y = 20;

printAdd(x, y);

const p = 30;

foo(x);
bar(y);
```

**Are you willing to submit a pull request to implement this change?**
Yeah

**Comment by mdjermanovic** [comment]
Hi @maluramichael, thanks for the issue!

This rule doesn't have a StatementType specifically for statements that consists of a function call, but they fall under `"expression"` so you can configure the rule to enforce a blank line between variable declarations and expression statements (incl. function calls) like this:

```
{ blankLine: "always", prev: ["const", "let", "var"], next: "expression" },
{ blankLine: "always", prev: "expression", next: ["const", "let", "var"] }
```

[Online Demo link](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHBhZGRpbmctbGluZS1iZXR3ZWVuLXN0YXRlbWVudHM6IFtcbiAgICBcImVycm9yXCIsXG4gICAgeyBibGFua0xpbmU6IFwiYWx3YXlzXCIsIHByZXY6IFtcImNvbnN0XCIsIFwibGV0XCIsIFwidmFyXCJdLCBuZXh0OiBcImV4cHJlc3Npb25cIiB9LFxuICAgIHsgYmxhbmtMaW5lOiBcImFsd2F5c1wiLCBwcmV2OiBcImV4cHJlc3Npb25cIiwgbmV4dDogW1wiY29uc3RcIiwgXCJsZXRcIiwgXCJ2YXJcIl0gfVxuXSovXG5cbmNvbnN0IHggPSAxMDtcbmNvbnN0IHkgPSAyMDtcbmNvbnN0IHJlc3VsdCA9IGFkZCh4LCB5KTsgLy8gPC0gdGhhdHMgb2sgYmVjYXVzZSBpdCBpcyBhbHNvIGEgdmFyaWFibGUgZGVjbGFyYXRpb25cbmNvbnN0IHAgPSAzMDtcblxuY29uc3QgeDEgPSAxMDtcbmNvbnN0IHkxID0gMjA7XG5wcmludEFkZCh4LCB5KTsgLy8gPC0gdGhpcyB3b3VsZCBiZSB3cm9uZyBiZWNhdXNlIHRoZSBkZWNsYXJhdGlvbiBvZiB5IGlzIGZvbGxvd2VkIGJ5IGEgY2FsbCB0byBwcmludEFkZFxuY29uc3QgcDEgPSAzMDsgLy8gPC0gd291bGQgYmUgYWxzbyB3cm9uZ1xuZm9vKHgpO1xuYmFyKHkpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by maluramichael** [comment]
@mdjermanovic thank you for the quick answer. I think i tried that but had problems somewhere else. For example this code here should not add a space.

```js
const element            = document.createElement('a');
element.href             = destination;
element.download         = name;
element.style.visibility = 'hidden';
```

But with this setting it does.

```js
{ blankLine: "always", prev: ["const", "let", "var"], next: "expression" },
{ blankLine: "always", prev: "expression", next: ["const", "let", "var"] }
```

Is there a finer way to differentiate between variable declaration (`const|let|var a = 10;` ) and just a variable assignment like `a=10;`

I tried to add a custom statement type for a quick test.

```js
"function-call": {
    test: (node, sourceCode) =>
              node.type === "ExpressionStatement" &&
              node.expression.type === "CallExpression"
},
```

This works ok for my simple example here.

```js
const x = 10;
const y = 20;
printAdd(x, y);
const p = 30;
```

But not here.

```js
const element            = document.createElement('a');
element.href             = destination;
element.download         = name;
element.style.visibility = 'hidden';
document.body.appendChild(element);
element.click();
document.body.removeChild(element);
```

Maybe because document.body.appendChild is a chained call?

**Comment by nzakas** [comment]
Because this is a stylistic rule, and we aren’t making any further changes to rules like this, converting to a discussion.
